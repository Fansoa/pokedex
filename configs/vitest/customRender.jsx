import { render } from "@testing-library/react";
import IntegrationProvider from "@/configs/vitest/testUtils";

const renderWithIntegrationProvider = (children) =>
  render(<IntegrationProvider>{children}</IntegrationProvider>);

export default renderWithIntegrationProvider;
