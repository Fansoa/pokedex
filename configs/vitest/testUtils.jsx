import { ThemeProvider } from "@emotion/react";
import PropTypes from "prop-types";
import theme from "@/src/theme/theme";

const IntegrationProvider = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

IntegrationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default IntegrationProvider;
