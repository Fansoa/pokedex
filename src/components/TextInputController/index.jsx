import PropTypes from "prop-types";
import { Controller, useFormContext } from "react-hook-form";

const TextInputController = ({ name, ...props }) => {
  const { control } = useFormContext();
  return (
    <Controller
      control={control}
      name={name}
      render={({ field: { onChange, value }, formState: { errors } }) => (
        <>
          <input
            type="text"
            onChange={onChange}
            value={value || ""}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
          />
          {errors?.[name]?.message}
        </>
      )}
    />
  );
};

TextInputController.propTypes = {
  name: PropTypes.string.isRequired,
};

export default TextInputController;
