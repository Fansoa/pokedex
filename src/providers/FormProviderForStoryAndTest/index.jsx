import { FormProvider, useForm } from "react-hook-form";
import PropTypes from "prop-types";

const FormProviderForStoryAndTest = ({ children }) => {
  const methods = useForm();
  return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    <FormProvider {...methods}>{children}</FormProvider>
  );
};

FormProviderForStoryAndTest.propTypes = {
  children: PropTypes.node,
};

FormProviderForStoryAndTest.defaultProps = {
  children: null,
};

export default FormProviderForStoryAndTest;
