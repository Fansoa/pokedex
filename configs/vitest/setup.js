import { expect, afterEach } from "vitest";
import { cleanup } from "@testing-library/react";
import * as matchers from "@testing-library/jest-dom/matchers";
import { createSerializer } from "@emotion/jest";

expect.extend(matchers);
expect.addSnapshotSerializer(createSerializer());

afterEach(() => {
  cleanup();
});
