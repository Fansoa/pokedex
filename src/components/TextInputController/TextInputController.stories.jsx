import TextInputController from "@/src/components/TextInputController";
import FormProviderForStoryAndTest from "@/src/providers/FormProviderForStoryAndTest";

export default {
  title: "Components/TextInputController",
  component: TextInputController,
  decorators: [
    (Story) => (
      <FormProviderForStoryAndTest>
        <Story />
      </FormProviderForStoryAndTest>
    ),
  ],
};

export const TextInputControllerStory = {
  name: "Default",
  args: {
    name: "inputName",
  },
};

export const TextInputControllerStoryWithLabel = {
  name: "With label",
  args: {
    name: "inputName",
    label: "Pokemon name or id",
  },
};

export const TextInputControllerStoryWithErrorMessage = {
  name: "With error message",
  args: {
    name: "inputName",
    label: "Pokemon name or id",
    errorMessage: "Message d'erreur",
  },
};
