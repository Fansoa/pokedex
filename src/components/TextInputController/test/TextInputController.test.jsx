import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { expect } from "vitest";
import renderWithIntegrationProvider from "@/configs/vitest/customRender";
import TextInputController from "@/src/components/TextInputController";
import { act } from "react-dom/test-utils";
import FormProviderForStoryAndTest from "@/src/providers/FormProviderForStoryAndTest";

describe("TestComponent", () => {
  let container;

  it("TextInput value change when user enters text", async () => {
    container = renderWithIntegrationProvider(
      <FormProviderForStoryAndTest>
        <TextInputController name="inputName" />
      </FormProviderForStoryAndTest>,
    ).container;
    const user = userEvent.setup();
    const input = screen.getByRole("textbox");
    expect(input).toBeInTheDocument();
    await act(async () => {
      await user.type(input, "message here!");
    });
    expect(input).toHaveValue("message here!");
    expect(container).toMatchSnapshot();
  });
});
