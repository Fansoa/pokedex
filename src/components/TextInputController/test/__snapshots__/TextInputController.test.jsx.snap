// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TestComponent > TextInput value change when user enters text 1`] = `
<div>
  <input
    type="text"
    value="message here!"
  />
</div>
`;
